processor/internal/user_transformer/user_transformer.go

```diff
@@ -0,0 +1,371 @@
+package user_transformer
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	jsoniter "github.com/json-iterator/go"
+	"github.com/samber/lo"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	"github.com/rudderlabs/rudder-go-kit/stats"
+
+	transformerclient "github.com/rudderlabs/rudder-server/internal/transformer-client"
+	"github.com/rudderlabs/rudder-server/processor/integrations"
+	"github.com/rudderlabs/rudder-server/processor/internal/transformer_utils"
+	"github.com/rudderlabs/rudder-server/processor/types"
+	"github.com/rudderlabs/rudder-server/utils/httputil"
+	reportingTypes "github.com/rudderlabs/rudder-server/utils/types"
+)
+
+var json = jsoniter.ConfigCompatibleWithStandardLibrary
+
+type UserTransformer struct {
+	config struct {
+		userTransformationURL      string
+		maxRetry                   config.ValueLoader[int]
+		failOnUserTransformTimeout config.ValueLoader[bool]
+		failOnError                config.ValueLoader[bool]
+		maxRetryBackoffInterval    config.ValueLoader[time.Duration]
+		timeoutDuration            time.Duration
+		collectInstanceLevelStats  bool
+		maxConcurrency             int
+	}
+	conf             *config.Config
+	log              logger.Logger
+	stat             stats.Stats
+	client           transformerclient.Client
+	guardConcurrency chan struct{}
+}
+
+type Opt func(*UserTransformer)
+
+func WithClient(client transformerclient.Client) Opt {
+	return func(s *UserTransformer) {
+		s.client = client
+	}
+}
+
+func (u *UserTransformer) SendRequest(ctx context.Context, clientEvents []types.TransformerEvent, batchSize int) types.Response {
```

Comment:
I believe transform is a more appropriate name.
Also, maybe this is a good opportunity to remove batchSize for the method and pass it as an option

```suggestion
func (u *UserTransformer) Transform(ctx context.Context, clientEvents []types.TransformerEvent, batchSize int) types.Response {
```